compiler/rustc_lint/src/tail_expr_drop_order.rs

```diff
@@ -0,0 +1,216 @@
+use std::mem::swap;
+
+use rustc_ast::UnOp;
+use rustc_hir::def::Res;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_hir::{Block, Expr, ExprKind, LetStmt, Pat, PatKind, QPath, StmtKind};
+use rustc_macros::LintDiagnostic;
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::Span;
+use tracing::debug;
+
+use crate::{LateContext, LateLintPass};
+
+declare_lint! {
+    /// ### What it does
+    /// Edition 2024 introduces a new rule with drop orders for values generated in tail expressions
+    /// of blocks.
+    /// Now the values will be dropped first, before the local variable bindings were dropped.
+    ///
+    /// This lint looks for those values generated at the tail expression location, that of type
+    /// with a significant `Drop` implementation, such as locks.
+    /// In case there are also local variables of type with significant `Drop` implementation as well,
+    /// this lint warns you of a potential transposition in the drop order.
+    /// Your discretion on the new drop order introduced by Edition 2024 is required.
+    ///
+    /// ### Why is this bad?
+    /// Values of type with significant `Drop` implementation has an ill-specified drop order that
+    /// come after those of local variables.
+    /// Edition 2024 moves them, so that they are dropped first before dropping local variables.
+    ///
+    /// ### Example
+    /// ```compile_fail,E0597
+    /// fn edition_2024() -> i32 {
+    ///     let mutex = std::sync::Mutex::new(vec![0]);
+    ///     mutex.lock().unwrap()[0]
+    /// }
+    /// ```
+    /// This lint only points out the issue with `mutex.lock()`, which will be dropped before `mutex` does.
+    /// No fix will be proposed.
+    /// However, the most probable fix is to hoist `mutex.lock()` into its own local variable binding.
+    /// ```no_run
+    /// fn edition_2024() -> i32 {
+    ///     let mutex = std::sync::Mutex::new(vec![0]);
+    ///     let guard = mutex.lock().unwrap();
+    ///     guard[0]
+    /// }
+    /// ```
```

Comment:
The doc comments for these `declare_lint!` declarations follow something of a template.  E.g.:

```rust
    /// The `tail_expr_drop_order` lint detects ...
    ///
    /// ### Example
    ///
    /// ```rust
    /// todo!()
    /// ```
    ///
    /// {{produces}}
    ///
    /// ### Explanation
    ///
    /// ...
```

Have a look at the others in e.g. [builtin.rs](https://github.com/spastorino/rust/blob/master/compiler/rustc_lint_defs/src/builtin.rs).

Reply:
I used the `rustc-clippy` templated. Edited to the new right template.